distsql/select_result.go

```diff
@@ -145,6 +147,28 @@ func (r *selectResult) NextRaw(ctx context.Context) (data []byte, err error) {
 // Next reads data to the chunk.
 func (r *selectResult) Next(ctx context.Context, chk *chunk.Chunk) error {
 	chk.Reset()
+	// Check the returned data is default/arrow format.
+	if r.selectResp == nil || (len(r.selectResp.RowBatchData) == 0 && r.respChkIdx == len(r.selectResp.Chunks)) {
+		err := r.getSelectResp()
+		if err != nil || r.selectResp == nil {
+			return err
+		}
+		if len(r.selectResp.RowBatchData) == 0 {
+			r.encodeType = tipb.EncodeType_TypeDefault
+			metrics.DistSQLDecodeTypeErrorCount.Inc()
+		}
+	}
+
+	switch r.encodeType {
+	case tipb.EncodeType_TypeDefault:
+		return r.readFromDefault(ctx, chk)
+	case tipb.EncodeType_TypeArrow:
+		return r.readFromArrow(ctx, chk)
+	}
+	panic("unsupported decode type")
+}
+
+func (r *selectResult) readFromDefault(ctx context.Context, chk *chunk.Chunk) error {
 	for !chk.IsFull() {
 		if r.selectResp == nil || r.respChkIdx == len(r.selectResp.Chunks) {
```

Comment:
no need to do this check again.

Reply:
Update.  `r.selectResp == nil` is unnecessary, but `r.respChkIdx == len(r.selectResp.Chunks)` is need.